core/arch/arm/plat-imx/pm/psci-suspend-imx7.S

```diff
@@ -0,0 +1,732 @@
+
+	.endm
+
+	.macro ddr_exit_retention
+
+	cmp	r5, #0x0
+	ldreq	r1, [r0, #PM_INFO_MX7_ANATOP_V_OFF]
+	ldrne	r1, [r0, #PM_INFO_MX7_ANATOP_P_OFF]
+	ldreq	r2, [r0, #PM_INFO_MX7_SRC_V_OFF]
+	ldrne	r2, [r0, #PM_INFO_MX7_SRC_P_OFF]
+	ldreq	r3, [r0, #PM_INFO_MX7_DDRC_V_OFF]
+	ldrne	r3, [r0, #PM_INFO_MX7_DDRC_P_OFF]
+	ldreq	r4, [r0, #PM_INFO_MX7_DDRC_PHY_V_OFF]
+	ldrne	r4, [r0, #PM_INFO_MX7_DDRC_PHY_P_OFF]
+	ldreq	r10, [r0, #PM_INFO_MX7_CCM_V_OFF]
+	ldrne	r10, [r0, #PM_INFO_MX7_CCM_P_OFF]
+	ldreq	r11, [r0, #PM_INFO_MX7_IOMUXC_GPR_V_OFF]
+	ldrne	r11, [r0, #PM_INFO_MX7_IOMUXC_GPR_P_OFF]
+
+	/* turn on ddr power */
+	ldr	r7, =BIT32(29)
+	str	r7, [r1, #ANADIG_SNVS_MISC_CTRL_CLR]
+
+	ldr	r6, =50
+	wait_delay
+
+	/* clear ddr_phy reset */
+	ldr	r6, =0x1000
+	ldr	r7, [r2, r6]
+	orr	r7, r7, #0x3
+	str	r7, [r2, r6]
+	ldr	r7, [r2, r6]
+	bic	r7, r7, #0x1
+	str	r7, [r2, r6]
+13:
+	ldr	r6, [r0, #PM_INFO_MX7_DDRC_REG_NUM_OFF]
+	ldr	r7, =PM_INFO_MX7_DDRC_REG_OFF
+	add	r7, r7, r0
+14:
+	ldr	r8, [r7], #0x4
+	ldr	r9, [r7], #0x4
+	str	r9, [r3, r8]
+	subs	r6, r6, #0x1
+	bne	14b
+	ldr	r7, =0x20
+	str	r7, [r3, #DDRC_PWRCTL]
+	ldr	r7, =0x0
+	str	r7, [r3, #DDRC_DFIMISC]
+
+	/* do PHY, clear ddr_phy reset */
+	ldr	r6, =0x1000
+	ldr	r7, [r2, r6]
+	bic	r7, r7, #0x2
+	str	r7, [r2, r6]
+
+	ldr	r7, [r1, #ANADIG_DIGPROG]
+	and	r7, r7, #0xff
+	cmp	r7, #0x11
+	bne	12f
+
+	/*
+	 * TKT262940:
+	 * System hang when press RST for DDR PAD is
+	 * in retention mode, fixed on TO1.1
+	 */
+	ldr	r7, [r11]
+	bic	r7, r7, #BIT32(27)
+	str	r7, [r11]
+	ldr	r7, [r11]
+	bic	r7, r7, #BIT32(29)
+	str	r7, [r11]
+12:
+	ldr	r7, =BIT32(30)
+	str	r7, [r1, #ANADIG_SNVS_MISC_CTRL_SET]
+
+	/* need to delay ~5mS */
+	ldr	r6, =0x100000
+	wait_delay
+
+	ldr	r6, [r0, #PM_INFO_MX7_DDRC_PHY_REG_NUM_OFF]
+	ldr	r7, =PM_INFO_MX7_DDRC_PHY_REG_OFF
+	add	r7, r7, r0
+
+15:
+	ldr	r8, [r7], #0x4
+	ldr	r9, [r7], #0x4
+	str	r9, [r4, r8]
+	subs	r6, r6, #0x1
+	bne	15b
+
+	ldr	r7, =0x0
+	add	r9, r10, #0x4000
+	str	r7, [r9, #0x130]
+
+	ldr	r7, =0x170
+	orr	r7, r7, #0x8
+	str	r7, [r11, #0x20]
+
+	ldr	r7, =0x2
+	add	r9, r10, #0x4000
+	str	r7, [r9, #0x130]
+
+	ldr	r7, =0xf
+	str	r7, [r4, #DDRPHY_LP_CON0]
+
+	/* wait until self-refresh mode entered */
+16:
+	ldr	r7, [r3, #DDRC_STAT]
+	and	r7, r7, #0x3
+	cmp	r7, #0x3
+	bne	16b
+	ldr	r7, =0x0
+	str	r7, [r3, #DDRC_SWCTL]
+	ldr	r7, =0x1
+	str	r7, [r3, #DDRC_DFIMISC]
+	ldr	r7, =0x1
+	str	r7, [r3, #DDRC_SWCTL]
+17:
+	ldr	r7, [r3, #DDRC_SWSTAT]
+	and	r7, r7, #0x1
+	cmp	r7, #0x1
+	bne	17b
+18:
+	ldr	r7, [r3, #DDRC_STAT]
+	and	r7, r7, #0x20
+	cmp	r7, #0x20
+	bne	18b
+
+	/* let DDR out of self-refresh */
+	ldr	r7, =0x0
+	str	r7, [r3, #DDRC_PWRCTL]
+19:
+	ldr	r7, [r3, #DDRC_STAT]
+	and	r7, r7, #0x30
+	cmp	r7, #0x0
+	bne	19b
+
+20:
+	ldr	r7, [r3, #DDRC_STAT]
+	and	r7, r7, #0x3
+	cmp	r7, #0x1
+	bne	20b
+
+	/* enable port */
+	ldr	r7, =0x1
+	str	r7, [r3, #DDRC_PCTRL_0]
+
+	/* enable auto self-refresh */
+	ldr	r7, [r3, #DDRC_PWRCTL]
+	orr	r7, r7, #(1 << 0)
+	str	r7, [r3, #DDRC_PWRCTL]
+
+	.endm
+
```

Comment:
Could we add a comment to state that the routine `imx7_suspend` is copied to coherent memory and executed from there without any dependency on code/data in tee memory (DDR).

Reply:
Have added comments in the beginning of this file:
```
+/*
 + * The code in this file is copied to coherent on-chip ram memory,
 + * without any dependency on code/data in tee memory(DDR).
 + */
```